pkg/pillar/cmd/volumemgr/handlepersist.go

```diff
@@ -0,0 +1,135 @@
+package volumemgr
+
+import (
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	log "github.com/sirupsen/logrus"
+	"time"
+)
+
+func lookupPersistImageStatus(ctx *volumemgrContext, objType string,
+	imageSha string) *types.PersistImageStatus {
+
+	if imageSha == "" {
+		return nil
+	}
+	pub := ctx.publication(types.PersistImageStatus{}, objType)
+	s, _ := pub.Get(imageSha)
+	if s == nil {
+		log.Infof("lookupPersistImageStatus(%s) not found for %s", imageSha, objType)
+		return nil
+	}
+	status := s.(types.PersistImageStatus)
+	return &status
+}
+
+func publishPersistImageStatus(ctx *volumemgrContext,
+	status *types.PersistImageStatus) {
+	log.Debugf("publishPersistImageStatus(%s, %s)",
+		status.ObjType, status.ImageSha256)
+
+	pub := ctx.publication(types.PersistImageStatus{}, status.ObjType)
+	key := status.Key()
+	pub.Publish(key, *status)
+}
+
+func unpublishPersistImageStatus(ctx *volumemgrContext,
+	status *types.PersistImageStatus) {
+
+	log.Debugf("publishPersistImageStatus(%s, %s)",
+		status.ObjType, status.ImageSha256)
+
+	pub := ctx.publication(types.PersistImageStatus{}, status.ObjType)
+	key := status.Key()
+	st, _ := pub.Get(key)
+	if st == nil {
+		log.Errorf("unpublishPersistImageStatus(%s) not found", key)
+		return
+	}
+	pub.Unpublish(key)
+}
+
+//AddOrRefCountPersistImageStatus calculates and updates refcount of a PersistImageStatus.
+//Creates a PersistImageStatus with the refcount if there is no PersistImageStatus found for the imageSha.
+func AddOrRefCountPersistImageStatus(ctx *volumemgrContext, name, objType, fileLocation, imageSha string, size int64) {
```

Comment:
This and its ReduceRefCountPersistImageStatus cousin have a few issues where I think the refcount can become incorrect.
First of all, you call ReduceRefCountPersistImageStatus even if you do not have a reference from that particular volumestatus to the PersistImageStatus. I think you need to add a HasPeristRef boolean to VolumeStatus so you can make sure you only reduce after you've done an add.

Second of all, AddOrRefCountPersistImageStatus doesn't do an add; it does a update based on all the VolumeStatus. That means we can not track who should decrement. 
The better pattern is to add a refcount from one particular VolumeStatus (and create with refcount=1 if the PersistImageStatus does not yet exist) and then have the caller set HasPersistRef in that VolumeStatus. Then we can ensure that the refcounts track correctly.

Reply:
Changed.